M1FExplained/Chapter10/Exercises/Exercise05.lean

```diff
@@ -26,31 +35,150 @@ lemma part_a (m n a : ℤ) (hmn : Int.gcd m n = 1) (hm : m ∣ a) (hn : n ∣ a)
     exact h₂ 
   match hc with 
   |⟨e, he⟩ => 
+  -- e is the number such that n * e = k₁. It also is the case that (m * n) * e = a, proving our statment. 
   use e
   calc
     a = k₁ * m := by exact hk₁
     _ = (n * e) * m := by rw [he]
     _ = m * n * e := by ring
 
-lemma part_b (m n : ℤ) (hmn : Int.gcd m n ≠ 1) : ∃ (a : ℤ), (m ∣ a) ∧ (n ∣ a) ∧ (¬(m * n) ∣ a) := by sorry
-  
+-- part b
+
+/- 
+Show that the conclusion of part a is false if m and n are not coprime (i.e, show that if  and n
+are not coprime, here exists an integer a such that m ∣ a and n ∣ a but ¬ m * n ∣ a)
+-/
+
+/- 
+Note that this statment of the question as in the book is incorrect. If we restrict m and n to be 
+non-zero, then the statment is true, so that is what we will prove.
+
+To help, we first define two helper lemmas as the solution is longwinded.
+-/
+
+-- helper_1 is a simple inequality involving absolute values.
+
+lemma helper_1 (m n M : ℤ) (hM : 1 < abs M) (hM' : M ∣ m * n) (hm : m ≠ 0) (hn : n ≠ 0) : abs (m * n / M)  < abs (m * n) := by
+match hM' with 
+|⟨k, hk⟩ => 
+have hkn0 : 0 < abs k := by {
+  rw [abs_pos]
+  intro h
+  rw [h] at hk
+  simp at hk
+  tauto
+}
+rw [hk, abs_mul, Int.mul_ediv_cancel_left]
+·calc
+  abs k = 1 * abs k := by norm_num
+  _ < abs M * abs k := by rel [hM]
+·intro h 
+ have hM0 : abs M = 0 := by simp; exact h
+ have : ¬ 1 < abs M := by rw [hM0]; norm_num
+ contradiction
+
+/- 
+helper_2 converts `Nat.not_dvd_of_pos_of_lt` to the integer case, casting to absolute values.
+It says that an integer n cannot divide any non-zero integer m which is smaller in magnitude.
+-/
+
+lemma helper_2 (m n : ℤ) (h1 : 0 < abs m) (h2 : abs m < abs n) : ¬ n ∣ m := by 
+  have h1' : 0 < Int.natAbs m := by {
+    rw [@Int.natAbs_pos]
+    exact Iff.mp abs_pos h1
+  }
+  have h2' : Int.natAbs m < Int.natAbs n := by {
+    rw [@Int.natAbs_lt_iff_sq_lt]
+    exact Iff.mpr sq_lt_sq h2
+  }
+  have := @Nat.not_dvd_of_pos_of_lt _ _ h1' h2'
+  simp at this
+  assumption
+
+-- Actual statment of part b:
+
+lemma part_b (m n : ℤ) (hm : m ≠ 0) (hn : n ≠ 0) (h2 : Int.gcd m n ≠ 1) : ∃ (a : ℤ), (m ∣ a) ∧ (n ∣ a) ∧ (¬(m * n) ∣ a) := by 
+  let M : ℤ := Int.gcd m n
+  /- Since m / M and n / M are both integers, our choice of a has both factors m and n. However m * n > a,
+  so cannot possibly divide it. -/
+  use (m * n / M)
+  apply And.intro
+  -- First prove that m ∣ a.
+  ·rw [show m * n / M = m * (n / M) by {
+    exact Int.mul_ediv_assoc m (show M ∣ n by exact Int.gcd_dvd_right m n)
+  }]
+   have : m ∣ m := by use 1; norm_num
```

Comment:
```suggestion
   have : m ∣ m := Int.dvd_refl m
```
This fact already has a name (and indeed you can guess what the name is if you understand the naming convention), so you probably don't need to make a new proof of it using `have`; just use `Int.dvd_refl m` whenever you need it.